pkg/proxy/iptables/proxier.go

```diff
@@ -0,0 +1,572 @@
+	if err != nil { // if we failed to get any rules
+		glog.Errorf("Failed to execute iptable-save, syncing all rules. (%s)", err.Error())
+	} else { // otherwise parse the output
+		existingChains = getChainLines(utiliptables.TableNAT, iptablesSaveRaw)
+	}
+
+	// for first line and chains
+	var chainsLines bytes.Buffer
+	// for the actual rules (which should be after the list of chains)
+	var rulesLines bytes.Buffer
+
+	// write table header
+	chainsLines.WriteString("*nat\n")
+
+	if chain, ok := existingChains[iptablesServicesChain]; ok {
+		chainsLines.WriteString(fmt.Sprintf("%s\n", chain))
+	} else {
+		chainsLines.WriteString(makeChainLine(iptablesServicesChain))
+	}
+
+	newHostChains := []utiliptables.Chain{}
+	newServiceChains := []utiliptables.Chain{}
+
+	//Build rules for services
+	for name, info := range proxier.serviceMap {
+		protocol := strings.ToLower((string)(info.portal.protocol))
+		// get chain name
+		svcChain := servicePortToServiceChain(name)
+		// Create chain
+		if chain, ok := existingChains[svcChain]; ok {
+			chainsLines.WriteString(fmt.Sprintf("%s\n", chain))
+		} else {
+			chainsLines.WriteString(makeChainLine(svcChain))
+		}
+		// get hosts and host-Chains
+		hosts := make([]string, 0)
+		hostChains := make([]utiliptables.Chain, 0)
+		for _, ep := range info.endpoints {
+			hosts = append(hosts, ep)
+			hostChains = append(hostChains, servicePortAndEndpointToServiceChain(name, ep))
+		}
+
+		// Ensure we know what chains to flush/remove next time we generate the rules
+		newHostChains = append(newHostChains, hostChains...)
+		newServiceChains = append(newServiceChains, svcChain)
+
+		// write chain and sticky session rule
+		for _, hostChain := range hostChains {
+			// Create chain
+			if chain, ok := existingChains[utiliptables.Chain(hostChain)]; ok {
+				chainsLines.WriteString(fmt.Sprintf("%s\n", chain))
+			} else {
+				chainsLines.WriteString(makeChainLine(hostChain))
+			}
+
+			// Sticky session
+			if info.sessionAffinityType == api.ServiceAffinityClientIP {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --rcheck --seconds %d --reap -j %s\n", svcChain, name.String(), hostChain, info.stickyMaxAgeMinutes*60, hostChain))
+			}
+		}
+
+		// write proxy/loadblanacing rules
+		n := len(hostChains)
+		for i, hostChain := range hostChains {
+			// Roughly round robin statistically if we have more than one host
+			if n > 1 {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m statistic --mode random --probability %f -j %s\n", svcChain, name.String(), 1.0/float64(n-i), hostChain))
+			} else {
+				rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -j %s\n", svcChain, name.String(), hostChain))
+			}
+			// proxy
+			rulesLines.WriteString(fmt.Sprintf("-A %s -m comment --comment \"%s\" -m recent --name %s --set -j DNAT -p %s --to-destination %s\n", hostChain, name.String(), hostChain, protocol, hosts[i]))
```

Comment:
We should only set `-m recent --name %s --set` if the affinity is clientIP.  Otherwise nothing